Clarify how secure auth works for root user

There was a recent chat in Gitter about connecting a Java client to a secure cluster with `root` (see below). In addition to adding code samples for secure connections to Build a Test App (), we need to clarify a few things in the docs:

- Based on the [code](https://github.com/cockroachdb/cockroach/blob/a54dd60af1a3d55ca3a7e1ff17309bc0d452ebb7/pkg/security/auth.go#L124), it looks like the `root` user cannot use password-based auth but rather MUST use client cert and key to authenticate. We need to double-check this and add to the docs. Also, the error message confused the user: `ERROR: user root must authenticate using a client certificate`. I think this error needs to clarify that client cert and key must be provided INSTEAD of password.

From Gitter:

```
christian rivasseau @christian-lefty 08:24
Hi does anyone have code snippets to connect to CDB using java and SSL ?

jseldess @jseldess 09:44
@christian-lefty, we have a docs issue to update the code samples on https://www.cockroachlabs.com/docs/build-a-test-app.html for secure connections: cockroachdb/docs#631. Hope to get to that early next year. In the meantime, perhaps one of the CockroachDB devs can help you out.

christian rivasseau @christian-lefty 09:57
@jseldess I think I m really close but I keep bumping into stupid errors
when I add a password I get Caused by: org.postgresql.util.PSQLException: ERROR: user root must authenticate using a client certificate
when I don't I get Caused by: org.postgresql.util.PSQLException: The server requested password-based authentication, but no password was provided.
I was just wondering if anyone ever got java + ssl working

jseldess @jseldess 10:04
@christian-lefty, hmm, our docs make it look like you need to pass the CA cert when authenticating with a password: https://www.cockroachlabs.com/docs/create-and-manage-users.html#secure-clusters-with-passwords.
That ERROR: user root must authenticate using a client certificate says something different, though. Are you passing the CA cert with the password?

christian rivasseau @christian-lefty 10:08
well I thought I was, I'm doing this
   try {
    // cockroach uses the postgreSQL driver.
    Class.forName("org.postgresql.Driver");
  } catch (ClassNotFoundException ex) {
    throw new IllegalStateException("please package postgresql jar in this app");
  }

  Properties props = new Properties();
  props.setProperty("ssl", "true");
  props.setProperty("charSet", "UTF-8");
  props.setProperty("sslcert", "config/keys/cockroach/client_user_root.cert");
  props.setProperty("sslkey", "config/keys/cockroach/client_user_root.key");
  props.setProperty("sslrootcert", "config/keys/cockroach/caroot.cert");
  props.setProperty("user", "root");
  props.setProperty("password", "**");

  // Connect to the database.
  conn = DriverManager.getConnection("jdbc:postgresql://online-cockroach-a-pub:26257/lefty?user=root&password=**",
      props);

jseldess @jseldess 10:12
So when you remove that password property and the password parameter in the connection string, you get The server requested password-based authentication, but no password was provided.?

christian rivasseau @christian-lefty 10:22
@jseldess OK so it appears that my Properties thing doesn't work apparently, when I put everything in the JDBC connection string it seems to send the cert
now I'm hitting Caused by: org.postgresql.util.PSQLException: Could not read SSL key file
Caused by: java.io.IOException: extra data given to DerValue constructor
at least this seems to imply that the key is being read

jseldess @jseldess 10:28
OK, glad there's a little progress at least  I'm actually a little confused about the expected behavior for root. I think root is special in that you must use the client cert and key to auth, not a password. But @asubiotto should be able to clarify.
Can you past your updated connection string?

christian rivasseau @christian-lefty 10:31
Yes I believe CLIENT cert is required
it's just the error message is smisleading but that's not a problem

Jordan Lewis @jordanlewis 10:32
Seems like you might have imported the certificate into your java keystore incorrectly: http://stackoverflow.com/questions/11621414/reading-an-x-509-certificate-with-java

christian rivasseau @christian-lefty 10:41
  updated string: conn = DriverManager.getConnection("jdbc:postgresql://online-cockroach-a-pub:26257/lefty?sslcert=%2Fhome%2Flefty%2Flefty%2Fconfig%2Fkeys%2Fcockroach%2Fclient_user_root.cert&sslkey=%2Fhome%2Flefty%2Flefty%2Fconfig%2Fkeys%2Fcockroach%2Fclient_user_root.key&sslmode=require&sslrootcert=%2Fhome%2Flefty%2Flefty%2Fconfig%2Fkeys%2Fcockroach%2Fcaroot.cert");
where the .cert and .key files are the ones created by cockroach cert

christian rivasseau @christian-lefty 10:46
@jordanlewis I haven't really imported the certificate
```

--- a/create-and-manage-users.md
+++ b/create-and-manage-users.md
@@ -56,7 +56,7 @@ Flag | Description
 `--host` | Database server host to connect to.<br/><br/>**Env Variable:** `COCKROACH_HOST`
 `--insecure` | Set this only if the cluster is insecure and running on multiple machines. <br><br>If the cluster is insecure and local, leave this out. If the cluster is secure, leave this out and set the `--ca-cert`, `--cert`, and `--key` flags. <br><br>**Env Variable:** `COCKROACH_INSECURE`
 `--key` | Path to the [client key](create-security-certificates.html) protecting the client certificate of the user *issuing the command* (not the user you're creating). This flag is required if the cluster is secure. <br/><br/>**Env Variable:** `COCKROACH_KEY`
-`--password` | Enable password authentication for the user; you will be prompted to enter the password on the command line.<br/><br/>[Find more detail about how CockroachDB handles passwords](#user-authentication).
+`--password` | Enable password authentication for the user; you will be prompted to enter the password on the command line.<br/><br/>You cannot set a password for the `root` user.<br/><br/>[Find more detail about how CockroachDB handles passwords](#user-authentication).
 `-p`, `--port` | Connect to the cluster on the specified port.<br/><br/>**Env Variable:** `COCKROACH_PORT` <br/>**Default**: `26257`
 `--pretty` | Format tables using ASCII. When not specified, table rows are printed as tab-separated values (TSV). <br/><br/>**Default**: `true`
 `--url` | Connect to the cluster on the provided URL, e.g., `postgresql://myuser@localhost:26257/mydb`. If left blank, the connection flags are used (`host`, `port`, `user`, `database`, `insecure`, `certs`). <br/><br/>**Env Variable:** `COCKROACH_URL`
@@ -67,9 +67,9 @@ Flag | Description
 Secure clusters require users to authenticate their access to databases and tables. CockroachDB offers two methods for this:
 
 - [Client certificate and key authentication](#secure-clusters-with-client-certificates), which is available to all users. To ensure the highest level of security, we recommend only using client certificate and key authentication.
-- [Password authentication](#secure-clusters-with-passwords), which is available only to users with passwords. To set a password for a user, include the `--password` flag in the `cockroach user set` command. <br/><br/>You can use this password to authenticate users without supplying their client certificate and key; however, we recommend instead using client certificate and key authentication whenever possible.
+- [Password authentication](#secure-clusters-with-passwords), which is available only to users who you've created passwords for. To set a password for a user, include the `--password` flag in the `cockroach user set` command. However, you *cannot* add password authentication to the `root` user. <br/><br/>You can use this password to authenticate users without supplying their client certificate and key; however, we recommend instead using client certificate and key authentication whenever possible.
 
-{{site.data.alerts.callout_info}}Insecure clusters do not support user authentication, but you can still create passwords for users through the <code>--password</code> flag.{{site.data.alerts.end}}
+{{site.data.alerts.callout_info}}Insecure clusters do not support user authentication, but you can still create passwords for users (besides <code>root</code>) through the <code>--password</code> flag.{{site.data.alerts.end}}
 
 ## Examples
 
@@ -129,6 +129,8 @@ $ cockroach user set jpointsman \
 ~~~
 
 After issuing this command, enter and confirm the user's new password at the command prompt.
+
+{{site.data.alerts.callout_danger}}You cannot add password authentication to the <code>root</code> user.{{site.data.alerts.end}}
 
 ### List All Users
 

--- a/create-user.md
+++ b/create-user.md
@@ -41,9 +41,9 @@ table td:first-child {
 Secure clusters require users to authenticate their access to databases and tables. CockroachDB offers two methods for this:
 
 - [Client certificate and key authentication](#secure-clusters-with-client-certificates), which is available to all users. To ensure the highest level of security, we recommend only using client certificate and key authentication.
-- [Password authentication](#secure-clusters-with-passwords), which is available only to users with passwords. To create a user with a password, use the `WITH PASSWORD` clause of `CREATE USER`. <br/><br/>You can use this password to authenticate users without supplying their client certificate and key; however, we recommend instead using client certificate and key authentication whenever possible. <br/><br/>To add a password to an existing user, use the [`cockroach user` command](create-and-manage-users.html#update-a-users-password).
+- [Password authentication](#secure-clusters-with-passwords), which is available only to users who you've created passwords for. To create a user with a password, use the `WITH PASSWORD` clause of `CREATE USER`. <br/><br/>You can use this password to authenticate users without supplying their client certificate and key; however, we recommend instead using client certificate and key authentication whenever possible. <br/><br/>To add a password to an existing user, use the [`cockroach user` command](create-and-manage-users.html#update-a-users-password).
 
-{{site.data.alerts.callout_info}}Insecure clusters do not support user authentication, but you can still create passwords for users through the <code>WITH PASSWORD</code> clause.{{site.data.alerts.end}}
+{{site.data.alerts.callout_info}}Insecure clusters do not support user authentication, but you can still create passwords for users (besides <code>root</code>) through the <code>WITH PASSWORD</code> clause.{{site.data.alerts.end}}
 
 ## Examples
 

